src/multirust-cli/self_update.rs

```diff
@@ -0,0 +1,1000 @@
+    // of CARGO_HOME.
+    let numbah: u32 = rand::random();
+    let gc_exe = work_path.join(&format!("multirust-gc-{:x}.exe", numbah));
+
+    use winapi::{FILE_SHARE_DELETE,
+                 FILE_SHARE_READ, INVALID_HANDLE_VALUE,
+                 FILE_FLAG_DELETE_ON_CLOSE, DWORD, SECURITY_ATTRIBUTES,
+                 STARTUPINFOW, PROCESS_INFORMATION, OPEN_EXISTING,
+                 CREATE_UNICODE_ENVIRONMENT, GENERIC_READ};
+    use kernel32::{CreateFileW, CloseHandle, CreateProcessW};
+    use std::os::windows::ffi::OsStrExt;
+    use std::ptr;
+    use std::thread;
+    use std::time::Duration;
+    use std::io;
+    use std::mem;
+
+    unsafe {
+        // Copy multirust (probably this process's exe) to the gc exe
+        try!(utils::copy_file(multirust_path, &gc_exe));
+
+        let mut gc_exe_win: Vec<_> = gc_exe.as_os_str().encode_wide().collect();
+        gc_exe_win.push(0);
+
+        // Open an inheritable handle to the gc exe marked
+        // FILE_FLAG_DELETE_ON_CLOSE. This will be inherited
+        // by subsequent processes.
+        let mut sa = mem::zeroed::<SECURITY_ATTRIBUTES>();
+        sa.nLength = mem::size_of::<SECURITY_ATTRIBUTES>() as DWORD;
+        sa.bInheritHandle = 1;
+
+        let gc_handle = CreateFileW(gc_exe_win.as_ptr(),
+                                    GENERIC_READ,
+                                    FILE_SHARE_READ | FILE_SHARE_DELETE,
+                                    &mut sa,
+                                    OPEN_EXISTING,
+                                    FILE_FLAG_DELETE_ON_CLOSE,
+                                    ptr::null_mut());
+
+        if gc_handle == INVALID_HANDLE_VALUE {
+            let err = io::Error::last_os_error();
+            return Err(Error::WindowsUninstallMadness(err));
+        }
+
+        defer!{{ let _ = CloseHandle(gc_handle); }}
+
+        let mut si = mem::zeroed::<STARTUPINFOW>();
+        si.cb = mem::size_of::<STARTUPINFOW>() as DWORD;
+        let mut pi = mem::zeroed::<PROCESS_INFORMATION>();
+
+        // Create the gc exe process, which will inherit a handle to
+        // its own file. It will wait for this process to exit then
+        // delete CARGO_HOME.
+        let r = CreateProcessW(gc_exe_win.as_ptr(),
+                               ptr::null_mut(),
+                               ptr::null_mut(), ptr::null_mut(),
+                               1, // inherit handles
+                               CREATE_UNICODE_ENVIRONMENT,
+                               ptr::null_mut(), ptr::null(),
+                               &mut si, &mut pi);
+
+        if r == 0 {
+            let err = io::Error::last_os_error();
+            return Err(Error::WindowsUninstallMadness(err));
+        }
+
+        // Sleep long enough to let the OS open the other process and increase
+        // its file handle count so we don't delete it before its run
+        thread::sleep(Duration::from_millis(100));
```

Comment:
holy cow, surely windows is better than this?


The SO answer at least didn't have this, but I guess your testing brought up that it should probably exist?